### PR TITLE
Update cerificate heightcheck to account for invalid BLS key

### LINKED ISSUE #8381: block doesn't get certified with `useInvalidBLSKeys`
### Expected behavior
block should get certified if BLS keys are registered during the init phase even if `useInvalidBLSKeys` are enabled.

### Actual behavior
block doesn't get certified forever.
Reason is
* When `useInvalidBLSKeys` are used, BLS keys will be registered during the init rounds (~309 blocks by default)
* at 309, engine will store new BFT properties with valid BLS keys as height 310. (this is because the new validators are active from height 310)
* aggregateCommit.height cannot be larger than `heightNextBFTParameters - 1`. In this case, `heightNextBFTParameters` is `310`, so at maximum `309` must be certified. However, BFTParameters for height `309` will not be created because there is no valid BLSkey.

### Steps to reproduce
Use `useINvalidBLSKeys` in the PoS module and register keys during the init rounds.
after around 1000 blocks, there should be aggregateCommits.heght at `0` where it should certify earlier block

### Which version(s) does this affect? (Environment, OS, etc...)
6.0.0-alpha.21

### Additional Information
LIP is updated with https://github.com/LiskHQ/lips/pull/334

### PR DESCRIPTION
**Pull request by shuse2**
### What was the problem?

This PR resolves #8381 

### How was it solved?

- Update config to include `minCertifyHeight`. This value should be minimal height key to be considered for the certification
- Update commit pool to take the height in the check

### How was it tested?

- Add unit test
- Link with core devnet and check if block will be certified after init rounds. (validator needs to be registered in the init round)

### CODE REVIEW

File: framework/src/engine/consensus/certificate_generation/commit_pool.ts [hunk 1]
```diff
@@ -308,6 +312,7 @@ export class CommitPool {
 				maxHeightCertified + 1,
 			);
 			nextHeight = Math.min(heightNextBFTParameters - 1, maxHeightPrecommitted);
+			nextHeight = Math.max(nextHeight, this._minCertifyHeight - 1);
```
Comment (ricott1) [inline framework/src/engine/consensus/certificate_generation/commit_pool.ts:315]: Why do we do it differently here? Couldn't we add

`heightNextBFTParameters = Math.max(heightNextBFTParameters, this._minCertifyHeight);`

after line 313 instead for consistency?

**Review by ricott1 — CHANGES_REQUESTED** [review]
I left a comment.

File: framework/src/engine/consensus/certificate_generation/commit_pool.ts (hunk 1 above)
Comment (shuse2) [reply framework/src/engine/consensus/certificate_generation/commit_pool.ts:315]: Assuming the `minCertrifyHeight` is one of the nextBFTParameters, i thought it's make sense because in L314, it is taking the min of `heightNextBFTParameters - 1, maxHeightPrecommitted`

Comment (ricott1) [reply framework/src/engine/consensus/certificate_generation/commit_pool.ts:315]: not sure I understand. The `nextHeight` calculated there should still be valid, I thought the problem was selecting the correct height of the 'first' BFT parameters.

Comment (gkoumout) [inline framework/src/engine/consensus/certificate_generation/commit_pool.ts:315]: I don't see why we use -1 here. 

My understanding is that `minCertifyHeight` we want to satisfy that  `getBFTParameters(minCertifyHeight)` would have the bls keys registered. So, it should be a first block of a round. For example, if all validators registered BLS keys during initRounds, it could be the first block of hybrid phase. 

If that's correct, then in order to produce a certificate, you need to have height at least `minCertifyHeight`, so I think the -1 should be dropped.

Comment (gkoumout) [inline framework/src/engine/consensus/certificate_generation/commit_pool.ts:315]: Moreover, I think there is a simpler and cleaner way to do this, correct me if I am wrong.

Don't make any change here, keep `selectAggregateCommit` same as before until line 323. And then just replace the while condition by:

`while (nextHeight > Math.max(maxHeightCertified, minCertifyHeight -1))`

This if `nextHeigh` is smaller than `minCertifyHeight`, then it will not even enter the while loop, since there is no point to do that.

**Review by gkoumout — CHANGES_REQUESTED** [review]
Left a couple of comments

File: framework/src/engine/consensus/certificate_generation/commit_pool.ts [hunk 2]
```diff
@@ -210,11 +212,13 @@ export class CommitPool {
 		}
 
 		try {
-			const heightNextBFTParameters = await this._bftMethod.getNextHeightBFTParameters(
+			let heightNextBFTParameters = await this._bftMethod.getNextHeightBFTParameters(
 				stateStore,
 				maxHeightCertified + 1,
 			);
 
+			heightNextBFTParameters = Math.max(heightNextBFTParameters, this._minCertifyHeight);
```
Comment (gkoumout) [inline framework/src/engine/consensus/certificate_generation/commit_pool.ts:220]: Similarly to my comment in `selectAggregateCommit`, I would revert this change here and just modify the check of line 206 

`if aggregateCommit.height <= Math.max(maxHeightCertified, minCertifyHeight -1)`

**Review by ricott1 — DISMISSED** [review]

File: framework/src/schema/application_config_schema.ts [hunk 3]
```diff
@@ -255,6 +255,11 @@ export const applicationConfigSchema = {
 					maximum: 30 * 1024, // Kilo Bytes
 					description: 'Maximum number of transactions allowed per block',
 				},
+				minimumCertifyHeight: {
+					type: 'integer',
+					minimum: 0,
```
Comment (ishantiw) [inline framework/src/schema/application_config_schema.ts:260]: Will it work for minimum value zero? as we give default value as 1

**Review by ishantiw — APPROVED** [review]
Approving from my end, one question though

File: framework/src/schema/application_config_schema.ts (hunk 3 above)
Comment (shuse2) [reply framework/src/schema/application_config_schema.ts:260]: ah good point, i should change this to `1`

File: framework/test/unit/engine/consensus/certificate_generation/commit_pool.spec.ts [hunk 4]
```diff
@@ -1380,5 +1410,19 @@ describe('CommitPool', () => {
 				certificateSignature: Buffer.alloc(0),
 			});
 		});
+
+		it('should certify the minCertifyHeight even when heightNextBFTParameters is lower', async () => {
+			// Arrange
+			bftMethod.getNextHeightBFTParameters.mockResolvedValue(309);
+			(commitPool['_minCertifyHeight'] as any) = heightNextBFTParameters - 1;
```
Comment (janhack) [inline framework/test/unit/engine/consensus/certificate_generation/commit_pool.spec.ts:1417]: Just to confirm, here we set `_minCertifyHeight` to 1052 and then expect the selected aggregate commit to have height 1052 and not 1051? This is correct, we could maybe additionally also test that if `_minCertifyHeight == 1051` then we get `singleCommit1`, i.e., the selected aggregate commit has height 1051.

**Review by janhack — APPROVED** [review]
The PR looks good. I just added one suggestion for an additional test.

Apart from that I was wondering if in the P2P gossip, we should discard any single commits with height `< MIN_CERTIFICATE_HEIGHT` as those would anyways not be selected. This is just a small optimization, but should not be necessary for the correctness of the protocol. If we do that then we should also update the LIP.

File: framework/test/unit/engine/consensus/certificate_generation/commit_pool.spec.ts [hunk 5]
```diff
@@ -1380,5 +1410,19 @@ describe('CommitPool', () => {
 				certificateSignature: Buffer.alloc(0),
 			});
 		});
+
+		it('should certify the minCertifyHeight even when heightNextBFTParameters is lower', async () => {
```
Comment (gkoumout) [inline framework/test/unit/engine/consensus/certificate_generation/commit_pool.spec.ts:1414]: I am not sure I understand what you mean here, since you defined 

`heightNextBFTParameters =  Math.max(heightNextBFTParameters, this._minCertifyHeight + 1)`

So I guess you mean "should certify the minCertifyHeight even when getNextHeightBFTParameters returns a lower value than `MIN_CERTIFICATE_HEIGHT + 1`". 

If this is the case, I would suggest to be careful and try both cases:

i ) `getNextHeightBFTParameters < MIN_CERTIFICATE_HEIGHT`
ii) `getNextHeightBFTParameters = MIN_CERTIFICATE_HEIGHT` (should accept again and this is the important case where the previous version had the problem of rejection)

Comment (gkoumout) [reply framework/test/unit/engine/consensus/certificate_generation/commit_pool.spec.ts:1414]: Not sure if the comment of @janhack covers the same case.

**Review by gkoumout — APPROVED** [review]
Left just one suggestion for an extra unit test where the case was not super clear to me.

File: framework/src/engine/consensus/certificate_generation/commit_pool.ts [hunk 6]
```diff
@@ -209,11 +211,17 @@ export class CommitPool {
 			return false;
 		}
 
+		// The heights of aggregate commits must be at least MIN_CERTIFY_HEIGHT.
```
Comment (ishantiw) [inline framework/src/engine/consensus/certificate_generation/commit_pool.ts:214]:
```suggestion
		// The heights of aggregate commits must be greater than or equal to MIN_CERTIFY_HEIGHT.
```

**Review by ishantiw — APPROVED** [review]

**Review by gkoumout — APPROVED** [review]

File: framework/test/unit/engine/consensus/certificate_generation/commit_pool.spec.ts (hunk 4 above)
Comment (janhack) [reply framework/test/unit/engine/consensus/certificate_generation/commit_pool.spec.ts:1445]: Isn't the test in line 1428 and line 1442 now the same? I could not spot the difference.

**Review by ishantiw — DISMISSED** [review]

File: framework/test/unit/engine/consensus/certificate_generation/commit_pool.spec.ts (hunk 5 above)
Comment (shuse2) [reply framework/test/unit/engine/consensus/certificate_generation/commit_pool.spec.ts]: Added the tests 👍

**Review by ishantiw — APPROVED** [review]

**Review by gkoumout — APPROVED** [review]

File: framework/test/unit/engine/consensus/certificate_generation/commit_pool.spec.ts (hunk 4 above)
Comment (janhack) [reply framework/test/unit/engine/consensus/certificate_generation/commit_pool.spec.ts:1445]: Looks good now.
